### PR TITLE
✨ Add support for optional patches in ClusterClass

### LINKED ISSUE #5566: Allow patches to be selectively applied
**User Story**

As an operator I would like to create ClusterClass's with a rich set of customizations that can selectively applied by a user who chooses to use them. 

**Detailed Description**

This is a formal request for the feature that has been mentioned in the [ClusterClass and Managed Topologies](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/202105256-cluster-class-and-managed-topologies.md#behavior-with-patches)
```
Note: We are considering to add a field to allow defining optional patches. 
This would allow adding optional patches in the ClusterClass and then activating them on a per-cluster basis via opt-in.
```

Just like a optional fields and default values improves the reusability of a type definition (by giving flexibility to what and how much information need to be supplied to initialize an object of said type), providing control on which of the ClusterClass's patches are applied will allow the consolidation of said patches into a smaller number of reusable ClusterClass's.
Without this functionality, the number of actual ClusterClass's will grow exponentially as the number of patches associated with a ClusterClass's that may or may not be needed under actual usage scenarios increases.

One way to introduce this functionality without complicating the actual patching action would be to introduce an additional field in the ClusterClassPatch or PatchDefinition. At its simplest, the field can be optionally set with the name of a variable, the value of which will determine whether the associated patch will be applied.

Two main use cases I expect to come up quite often:

1. Apply patch only if variable(s) are set. Often patch is only required if certain variables are set ("if proxy setting is set, patch KCP"; "if autoscaler is enabled, patch MachineDeployment with scaling limits", etc) 
2. Using a boolean ClusterClassVariable, semantically opaque to the patch engine, to control if a patch is applied. This is a generalization of 1. and gives the client an additional level of control on whether a patch is relevant. (e.g Perhaps the computation of the value of this variable is not straightforward and the client is now free to introduce whatever logic to arrive at the this boolean value)

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind feature
/area topology

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

This PR implements optional patches as discussed in https://github.com/kubernetes-sigs/cluster-api/issues/5566

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5566

### CODE REVIEW

**Comment by sbueringer** [comment]
Rebased on top of main, easier to see the diff now.

**Comment by sbueringer** [comment]
/test pull-cluster-api-verify-main 
looks like kustomize download flake

**Comment by fabriziopandini** [comment]
/retitle Add support for optional patches in ClusterClass
/lgtm

@vincepri

**Comment by vincepri** [comment]
/retitle ✨ Add support for optional patches in ClusterClass

File: controllers/topology/internal/extensions/patches/inline/json_patch_generator.go [hunk 1]
```diff
@@ -132,6 +142,22 @@ func templateMatchesSelector(templateRef *api.TemplateRef, selector clusterv1.Pa
...
+	}
+
+	// Rendered template.
+	value, err := renderValueTemplate(*enabledIf, variables)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to calculate value for enabledIf")
+	}
+
+	// Patch is enabled if the rendered template value is `true`.
+	return bytes.Equal(value.Raw, []byte(`true`)), nil
```
Comment (vincepri) [inline controllers/topology/internal/extensions/patches/inline/json_patch_generator.go:158]: Can we unmarshal into a boolean instead? Or does `value` already handle cases like there is a leading newline, etc?

Comment (sbueringer) [reply controllers/topology/internal/extensions/patches/inline/json_patch_generator.go:159]: We are using `yaml.Unmarshal` in `renderValueTemplate` which is able to handle leading and trailing new lines.

I pushed 2 new unit tests to also cover those cases (named something with `(even with leading and trailing new line)`)

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**